pkg/server/controller.go

```diff
@@ -145,6 +282,264 @@ type indexPage struct {
+	type callbackResponse struct {
+		ID            string
+		Email         string
+		VerifiedEmail bool
+		Picture       string
+	}
+
+	var userProfile callbackResponse
+	err = json.NewDecoder(resp.Body).Decode(&userProfile)
+	if err != nil {
+		return
+	}
+
+	name = userProfile.Email
+	return
+}
+
+func (ctrl *Controller) decodeGithubCallbackResponse(resp *http.Response) (name string, err error) {
+	type callbackResponse struct {
+		ID        int64
+		Email     string
+		Login     string
+		AvatarURL string
+	}
+
+	var userProfile callbackResponse
+	err = json.NewDecoder(resp.Body).Decode(&userProfile)
+	if err != nil {
+		return
+	}
+
+	name = userProfile.Login
+	return
+}
+
+func (ctrl *Controller) decodeGitLabCallbackResponse(resp *http.Response) (name string, err error) {
+	type callbackResponse struct {
+		ID        int64
+		Email     string
+		Username  string
+		AvatarURL string
+	}
+
+	var userProfile callbackResponse
+	err = json.NewDecoder(resp.Body).Decode(&userProfile)
+	if err != nil {
+		return
+	}
+
+	name = userProfile.Username
+	return
+}
+
+func (ctrl *Controller) callbacRedirectkHandler(getAccountInfoURL string, oauthConf *oauth2.Config, decodeResponse func(*http.Response) (string, error)) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		cookie, err := r.Cookie(stateCookieName)
+		if err != nil {
+			ctrl.httpServer.ErrorLog.Printf("There seems to be problem with state cookie: %v", err)
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+
+		if cookie == nil {
+			ctrl.httpServer.ErrorLog.Printf("Missing state cookie")
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+
+		cookieState := cookie.Value
+		requestState := r.FormValue("state")
+
+		if requestState != cookieState {
+			ctrl.httpServer.ErrorLog.Printf("invalid oauth state, expected %v got %v", cookieState, requestState)
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+
+		code := r.FormValue("code")
+		if code == "" {
+			ctrl.httpServer.ErrorLog.Printf("Code not found")
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+
+		token, err := oauthConf.Exchange(oauth2.NoContext, code)
+		if err != nil {
+			ctrl.httpServer.ErrorLog.Printf("Exchanging auth code for token failed with %v ", err)
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+
+		client := oauthConf.Client(oauth2.NoContext, token)
+		resp, err := client.Get(getAccountInfoURL)
+		if err != nil {
+			ctrl.httpServer.ErrorLog.Printf("Failed to get oauth user info: %v", err)
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+		defer resp.Body.Close()
+
+		name, err := decodeResponse(resp)
+		if err != nil {
+			ctrl.httpServer.ErrorLog.Printf("Decoding response body failed: %v", err)
+			http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+			return
+		}
+
+		claims := jwt.MapClaims{
```

Comment:
I think a helper method (e.g, `newJwtToken() *jwt.Token`) is more appropriate here.